[EventDispatcher] swap arguments of dispatch() to allow registering events by FQCN

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

PR green and ready.
Bundles/packages that want to provide FC/BC compat with the previous way of using the dispatcher should use `LegacyEventDispatcherProxy::decorate()` (see example in the PR)

From UPGRADE files:

 EventDispatcher
---------------

 * The signature of the `EventDispatcherInterface::dispatch()` method should be updated to `dispatch($event, string $eventName = null)`, not doing so is deprecated

HttpKernel
----------

 * Renamed `FilterControllerArgumentsEvent` to `ControllerArgumentsEvent`
 * Renamed `FilterControllerEvent` to `ControllerEvent`
 * Renamed `FilterResponseEvent` to `ResponseEvent`
 * Renamed `GetResponseEvent` to `RequestEvent`
 * Renamed `GetResponseForControllerResultEvent` to `ViewEvent`
 * Renamed `GetResponseForExceptionEvent` to `ExceptionEvent`
 * Renamed `PostResponseEvent` to `TerminateEvent`

Security
---------

 * The `ListenerInterface` is deprecated, turn your listeners into callables instead.
 * The `Firewall::handleRequest()` method is deprecated, use `Firewall::callListeners()` instead.

nicolas-grekas committed:
[EventDispatcher] swap arguments of dispatch() to allow registering events by FQCN

OskarStark commented on UPGRADE-5.0.md:
```diff
@@ -63,6 +63,7 @@ EventDispatcher
 ---------------
 
  * The `TraceableEventDispatcherInterface` has been removed.
+ * The signature of the "EventDispatcherInterface::dispatch()" method has been updated to "dispatch($event, string $eventName = null)"
```
```suggestion
 * The signature of the `EventDispatcherInterface::dispatch()` method has been updated to `dispatch($event, string $eventName = null)`
```

OskarStark commented on src/Symfony/Component/EventDispatcher/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+4.3.0
+-----
+
+ * The signature of the "EventDispatcherInterface::dispatch()" method should be updated to "dispatch($event, string $eventName = null)", not doing so is deprecated
```
```suggestion
 * The signature of the `EventDispatcherInterface::dispatch()` method should be updated to `dispatch($event, string $eventName = null)`, not doing so is deprecated
```

lyrixx commented:
I have some questions

1. Why this change ?
2. if the goal is to simply the event dispatcher usage, why don't you do something like `$eventName = $eventName ?: get_class($event)` ?
3. What alias are used for ?

ostrolucky commented on src/Symfony/Component/EventDispatcher/EventDispatcherInterface.php:
```diff
@@ -23,15 +23,13 @@
      * @return Event
      */
-    public function dispatch($eventName, Event $event = null);
+    public function dispatch($event);
```
```suggestion
    public function dispatch($event/*, string $eventName = null */);
```

ostrolucky reviewed (changes requested):
Some test failures are caused by this change, please fix. Other than that, this is pretty clever, I like it.

nicolas-grekas commented on src/Symfony/Component/Workflow/WorkflowEvents.php:
```diff
@@ -0,0 +1,58 @@
+ * To learn more about how form events work check the documentation
+ * entry at {@link https://symfony.com/doc/current/workflow/usage.html#using-events}.
+ */
+final class WorkflowEvents
```
updated

nicolas-grekas commented:
Thanks for the early reviews. I still have a few edges to fix (as spotted by the CI).

> Why this change ?

two sides: allowing one to subscribe to events using FQCN:
`return [RequestEvent::class => 'onRequest'];`
and allowing to trigger your own events by FQCN and thus save ones from having to invent a name:
`$dispatcher->dispatch(new MyEvent());`

> if the goal is to simply the event dispatcher usage, why don't you do something like $eventName = $eventName ?: get_class($event) ?

that's what happens now :)

> What alias are used for ?

at least for BC, to map `RequestEvent::class` to `kernel.request`. BUT I updated the implementation to have the mapping happen at compile time, so that runtime alias resolution is reverted now.

lyrixx commented on UPGRADE-5.0.md:
```diff
@@ -63,6 +63,7 @@ EventDispatcher
 ---------------
 
  * The `TraceableEventDispatcherInterface` has been removed.
+ * The signature of the `EventDispatcherInterface::dispatch()` method has been updated to `dispatch($event, string $eventName = null)`
```
What about `dispatch(/* Event */ $event, string $eventName = null)` to enforce the fact the "new" signature will only accept an `Event` instance ?

In 6.0 we could update it right ?

nicolas-grekas replied:
No need: we'll be able to do this in 5.0, since PHP 7.2 will allow adding the type-hint to the interface without breaking BC.

lyrixx commented:
I prefer this new implementation. This new DX is better.

stof commented on src/Symfony/Component/EventDispatcher/LegacyEventDispatcherProxy.php:
```diff
@@ -0,0 +1,132 @@
+            }
+        }
+
+        return $this->dispatcher->dispatch($event, $eventName);
```
this looks weird to me. Shouldn't this call the inner dispatcher with the event name as first argument, as it is precisely wrapping a legacy one ?

ro0NL replied:
i dont we test the legacy api anymore, is it? Perhaps unit test `LegacyEventDispatcherProxy` at least :)

nicolas-grekas replied:
fixed, and tests added.

stof commented on src/Symfony/Component/Security/Guard/composer.json:
```diff
@@ -23,6 +23,9 @@
     },
+    "conflict": {
+        "symfony/event-dispatcher": "<4.2"
+    }
```
this missing comma must be added

stof commented:
The aliases being compile-time only mean that any code meant to work with components only and not FrameworkBundle cannot use them, and that Form aliases are useless (as these are *not* dispatched on the dispatcher configured by FrameworkBundle as each Form instance has its own EventDispatcher)

nicolas-grekas commented:
Idea: look at listeners bound to the interfaces of an event.

unkind commented:
I like this step a lot. This was one of my "favorite" drawback of the `EventDispatcher` which I used as one of the arguments to NOT using this component.

I can share some thoughts/ideas of my event handling experience, probably it may help to improve the component. I tried to not discuss it here because you are locked with BC policies. But if you want to change something...

I think you are aware of debates regarding using term "event" for this kind of service. In my opinion, Symfony provides hook dispatcher/manager, not event one. You can say that it's just matter of taste, but it has architecture impacts. You lose ability to make async event handlers without evidence that abstraction is very leaky. It was discussed here, for example: #23441. And regarding naming here, for instance: https://twitter.com/mathiasverraes/status/1013654989682741248. Probably, you will never change it, but it's very noticeable thing that comes in mind when we talk about this component.

This little "terminology disagreement" leads to 2 completely different concepts of EventDispatcher/Bus. "That" event bus:
1. has immutable events; changed event is just different event; altering event is ability to travel to parallel Universe or just fraud with bank account, it depends on context;
2. has event handlers in an isolated context: exceptions/errors thrown by them never come to the `EventBus` caller because event handlers are like destructors: execution time is undetermined (right now, 2 seconds later or 1 minute later if queue lags, etc.);
3. has no priorities on DI-level. This is awesome. You don't need to pick between -32000, 42 or 9223372036854775807 priorities: there is no determined order of event listeners, just chill. It just requires to make explicit chain of event handlers by raising new event from the `A` event handler to the `B`. Or by executing command which leads to the new events. Your processes in the system become more explicit when you see which action depend on which one. `priority: int` is implicit and unreliable way to build a control flow;
4. doesn't allow to "stop propagation"; event handlers are completely independent from each other; however, hooks probably should be aware of each other and make such decisions; "stop propagation" is incompatible with event handlers completely: probably, I just want to log event, who the heck you think you are to stop me from doing that (_looking at `ExceptionToErrorPageListener`_)?;
5. has event handlers with no knowledge about sync/async dispatching, because it doesn't matter.
6. decouples event initiator from the handlers: we don't need to think "should I check returned object or not?"; we just don't care about them, we don't even know whether they exist or not: full agnosticism.

Naming matters.

---

> Idea: look at listeners bound to the interfaces of an event.

I prefer to the following logic:
1. If class has the only listener, we just match by type (`GameWasStarted` for this specific event, `GameEvent` interface for all events of the `Game` aggregate, `Event` for broadcast-listener, etc.);
2. if class has multiple listeners, we try to allow the only route for any event, i.e. if class has `onStarted(GameWasStarted $event, Context $context)` and `onAny(Event $event)`, `GameWasStarted` will be passed in the first method only to avoid headaches for developer; but it works only if there is explicit inheritance dependency like `final class GameWasStarted implements Event`; if event implements 2 independent interfaces and type hint against them, he makes it on his own risk.

sstok commented:
@unkind I think you confuse two concepts here, the Symfony Event Dispatcher is based on the https://en.wikipedia.org/wiki/Mediator_pattern which allows "events" to be mutable, it's naming is definitely off :) changing this fundamental logic is a major BC break, instead Symfony would have to introduce another component or class that truly implements immutable events.

Async handling requires more work as you need to handle serialization and transportation which the Messenger component already solves nicely. Yes, you don't have priorities, but this feature could be added.

stof commented:
What @unkind describes here looks like the Messenger component :smile:

unkind commented:
> I think you confuse two concepts here

I'm not sure about this, I explicitly stated that

> This little "terminology disagreement" leads to 2 completely different concepts of EventDispatcher/Bus.

It was just a little rant about naming in the first place. Naming disregard is a problem.

OskarStark commented on UPGRADE-4.3.md:
```diff
@@ -1,6 +1,7 @@
 UPGRADE FROM 4.2 to 4.3
 =======================
 
+<<<<<<< HEAD
```
Should be removed

OskarStark commented on src/Symfony/Component/Workflow/composer.json:
```diff
@@ -26,11 +26,14 @@
         "symfony/validator": "~3.4|~4.0"
     },
+    "conflict": {
+        "symfony/event-dispatcher": "<4.3"
```
why is this needed? the minimum version is already `^4.3`

nicolas-grekas replied:
because require-dev doesn't apply to the root app

nicolas-grekas commented:
PR green and ready. From UPGRADE files:

 EventDispatcher
---------------

 * The signature of the `EventDispatcherInterface::dispatch()` method should be updated to `dispatch($event, string $eventName = null)`, not doing so is deprecated

HttpKernel
----------

 * Renamed `FilterControllerArgumentsEvent` to `ControllerArgumentsEvent`
 * Renamed `FilterControllerEvent` to `ControllerEvent`
 * Renamed `FilterResponseEvent` to `ResponseEvent`
 * Renamed `GetResponseEvent` to `RequestEvent`
 * Renamed `GetResponseForControllerResultEvent` to `ViewEvent`
 * Renamed `GetResponseForExceptionEvent` to `ExceptionEvent`
 * Renamed `PostResponseEvent` to `TerminateEvent`

Security
---------

 * The `ListenerInterface` is deprecated, turn your listeners into callables instead.
 * The `Firewall::handleRequest()` method is deprecated, use `Firewall::callListeners()` instead.

javiereguiluz reviewed (approved):
I like this a lot ... specially the current event renaming. Much better and easier to understand!

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml:
```diff
@@ -4,6 +4,37 @@
+            <parameter key="Symfony\Component\Form\SubmitEvent">form.submit</parameter>
+            <parameter key="Symfony\Component\Form\PostSubmitEvent">form.post_submit</parameter>
+            <parameter key="Symfony\Component\Form\PreSetDataEvent">form.pre_set_data</parameter>
+            <parameter key="Symfony\Component\Form\PostSetDataEvent">form.post_set_data</parameter>
```
The namespace is not right for form events (classes are in `Symfony\Component\Form\Event`).

nicolas-grekas replied:
fixed thanks

ostrolucky commented on src/Symfony/Component/EventDispatcher/ImmutableEventDispatcher.php:
```diff
@@ -22,15 +22,26 @@ class ImmutableEventDispatcher implements EventDispatcherInterface
+        $eventName = 1 < \func_num_args() ? \func_get_arg(1) : null;
+
+        if (\is_scalar($event)) {
+            // deprecated
```
This doesn't require proper deprecation message, like regular EventDispatcher class?

nicolas-grekas replied:
The deprecation was already triggered by the call in the constructor, so we can skip it here.

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml:
```diff
@@ -4,6 +4,37 @@
     xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
 
+    <parameters>
+        <parameter key="event_dispatcher.aliases" type="collection">
```
event_aliases for clarity?

nicolas-grekas replied:
updated

ro0NL commented on src/Symfony/Bundle/SecurityBundle/Debug/WrappedListener.php:
```diff
@@ -11,24 +11,29 @@
     private static $hasVarDumper;
 
-    public function __construct(ListenerInterface $listener)
+    public function __construct($listener)
```
`@param callable $listener`?

nicolas-grekas replied:
done

ro0NL commented on src/Symfony/Component/EventDispatcher/ImmutableEventDispatcher.php:
```diff
@@ -22,15 +22,26 @@ class ImmutableEventDispatcher implements EventDispatcherInterface
-        return $this->dispatcher->dispatch($eventName, $event);
+        $eventName = 1 < \func_num_args() ? \func_get_arg(1) : null;
+
+        if (\is_scalar($event)) {
```
`dispatch(...func_get_args())` for a while instead?

nicolas-grekas replied:
What do you mean? The decorated dispatcher is wrapped by `LegacyEventDispatcherProxy` si we're sure about the way it behaves.

ro0NL replied:
never mind, i thought the original dispatcher was always wrapped one

ro0NL commented on src/Symfony/Component/Security/Http/Tests/Firewall/ExceptionListenerTest.php:
```diff
@@ -190,7 +190,11 @@ private function createEvent(\Exception $exception, $kernel = null)
+            return new ExceptionEvent($kernel, Request::create('/'), HttpKernelInterface::MASTER_REQUEST, $exception);
+        }
+
+        return new ExceptionEvent($kernel, Request::create('/'), HttpKernelInterface::MASTER_REQUEST, $exception);
```
`GetResponseForExceptionEvent`?

nicolas-grekas replied:
fixed

fabpot commented:
Thank you @nicolas-grekas.

Tobion commented on src/Symfony/Component/Workflow/WorkflowEvents.php:
```diff
@@ -0,0 +1,58 @@
+namespace Symfony\Component\Workflow;
+
+/**
+ * To learn more about how form events work check the documentation
```
Copy paste error: form events

nicolas-grekas replied:
fixed in b8ab296adf thanks

Kocal commented:
If people are asking, this is how we define event listeners (that use event FQCN) in `services.yaml`:

```yaml
services:
    App\EventListener\MyEventListener:
        tags:
            - { name: kernel.event_listener, event: 'App\Event\MyEvent', method: onEvent }
```

```php
$eventDispatcher->dispatch(new App\Event\MyEvent());

patrick-mcdougle commented on src/Symfony/Component/EventDispatcher/LegacyEventDispatcherProxy.php:
```diff
@@ -0,0 +1,142 @@
+            }
+        }
+
+        $listeners = $this->getListeners($eventName);
+
+        foreach ($listeners as $listener) {
+            if ($event->isPropagationStopped()) {
+                break;
+            }
+            $listener($event, $eventName, $this);
+        }
+
+        return $event;
```
Old news, but shouldn't this just be a call to dispatch with the args swapped? What if there's custom logic in the decorated dispatcher's dispatch method that is no longer called?

patrick-mcdougle replied:
These lines should be replaced with `$this->dispatch($eventName, $event);`

patrick-mcdougle replied:
@nicolas-grekas @fabpot this is technically a BC break and should be fixed in the next release of 4 (assuming there is one)
